cordformation/src/main/kotlin/net/corda/plugins/Dockerform.kt

```diff
@@ -18,51 +18,80 @@ import javax.inject.Inject
+        private const val DEFAULT_SSH_PORT = 22022;
+        private val DEFAULT_DIRECTORY: Path = Paths.get("build", "docker")
 
-        private const val dockerComposeFileVersion = "3"
+        private const val COMPOSE_SPEC_VERSION = "3"
 
-        private val yamlOptions = DumperOptions().apply {
+        private val YAML_FORMAT_OPTIONS = DumperOptions().apply {
             indent = 2
             defaultFlowStyle = DumperOptions.FlowStyle.BLOCK
         }
-        private val yaml = Yaml(yamlOptions)
+
+        private val YAML_MAPPER = Yaml(YAML_FORMAT_OPTIONS)
     }
 
     private val directoryPath: Path = project.projectDir.toPath().resolve(directory)
 
-    @get:InputFile
-    val dockerComposePath: Path = directoryPath.resolve("docker-compose.yml")
+
+    val dockerComposePath: Path
+        @InputFile
+        get() {
+            val wantedPath = directoryPath.resolve("docker-compose.yml")
+            if (!Files.exists(wantedPath)) {
+                Files.createDirectories(wantedPath.parent)
+                Files.createFile(wantedPath)
+            }
+            return wantedPath
+        }
 
     /**
      * This task action will create and install the nodes based on the node configurations added.
      */
     @TaskAction
     fun build() {
-        project.logger.lifecycle("Running Cordform task")
+        project.logger.lifecycle("Running DockerForm task")
         initializeConfiguration()
-        nodes.forEach(Node::installDockerConfig)
+        nodes.forEach { it -> it.installDockerConfig(DEFAULT_SSH_PORT) }
         installCordaJar()
         nodes.forEach(Node::installDrivers)
+        generateKeystoreAndSignCordappJar()
+        generateExcludedWhitelist()
         bootstrapNetwork()
-        nodes.forEach(Node::installCordapps)
         nodes.forEach(Node::buildDocker)
 
 
         // Transform nodes path the absolute ones
-        val services = nodes.map { it.containerName to mapOf(
-                "build" to directoryPath.resolve(it.nodeDir.name).toAbsolutePath().toString(),
-                "ports" to listOf(it.rpcPort)) }.toMap()
+        val services = nodes.map {
+            val nodeBuildDir = directoryPath.resolve(it.nodeDir.name).toAbsolutePath().toString()
+
+            it.containerName to mapOf(
+                    "volumes" to listOf(
+                            "$nodeBuildDir/node.conf:/etc/corda/node.conf",
+                            "$nodeBuildDir/certificates:/opt/corda/certificates",
+                            "$nodeBuildDir/logs:/opt/corda/logs",
+                            "$nodeBuildDir/persistence:/opt/corda/persistence",
+                            "$nodeBuildDir/cordapps:/opt/corda/cordapps",
+                            "$nodeBuildDir/network-parameters:/opt/corda/network-parameters",
+                            "$nodeBuildDir/additional-node-infos:/opt/corda/additional-node-infos",
+                            "$nodeBuildDir/drivers:/opt/corda/drivers"
+
+                    ),
```

Comment:
@roastario does this need to also expose the `p2pAddress` port?

Also, if this doesn't bind the ports to the corresponding host ports, will the generated `node.conf` files be valid? (as `docker compose` chooses an ephemeral port)

Reply:
as the nodes will be within the same docker network, you do not need to expose the p2p ports. 

As for the rpc ports, I don't think we need to bind them to a specific port as this would mean that if two networks were started they would start clashing.

Reply:
Thanks for clarifying. Tested it manually, looks like it's all working.